### ISSUE TITLE
torch.nn.functional.glu is wrong ?

### ISSUE THREAD

**Issue by tanaka-jp** [pytorch/pytorch#16029]
I think that the torch.nn.functional.glu is not as same as the paper.
  _Language Modeling with Gated Convolutional Networks_
  _https://arxiv.org/abs/1612.08083_

The input E is not split in half along dim to form A and B.
In paper, 
  A = E∗W + b.
  B = E∗V + c.

Can anyone check it ?
Thanks.

**Comment by colesbury** [comment]
I'm pretty confident that it's the same as the paper. Which part in particular do you think differs?

(https://github.com/pytorch/fairseq/commits/master/fairseq/models/fconv.py is a usage example, which shares authors with the "Language Modeling with Gated Convolutional Networks" paper)

**Comment by tanaka-jp** [comment]
Maybe it is my misunderstanding.
But I think this description is wrong.
The input E is not splitted in half.
https://github.com/pytorch/pytorch/blob/fb68d813be3ccbab3cbf74008a11071b9a646c75/torch/nn/functional.py#L884


I didn't check the source code.

And I found another implementation of glu.
  https://github.com/jojonki/Gated-Convolutional-Networks/blob/master/gated_cnn.py

**Comment by colesbury** [comment]
It's split in both the PyTorch code and in the paper. See Figure 1, "Convolution".

**Comment by colesbury** [comment]
You're also confusing the input to the convolution or linear layer with the input to the gated linear unit. The input to the convolution (**E**) is not split. The input to the gated linear unit is split (that's **A** and **B**). GLU implements:

```
H0 = A⊗σ(B) 
```

Where **A** and **B** are the two halves of the input.

**Comment by tanaka-jp** [comment]
Thank you for your answer.

**Comment by villmow** [comment]
Why is this even splitted? Can't we provide A and B to GLU?

**Comment by leimao** [comment]
I came back to this GLU many years after I implemented it. The figure in the original paper is confusing, and there is no "split" for the inputs. Essentially you do two independent convolutions for the same input, get two outputs of the same shape, do sigmoid activation for one of the outputs, then multiply these two together. If the PyTorch `glu` is indeed doing splitting, then I think the programmer might have misunderstood the algorithm.

**Comment by harrygcoppock** [comment]
Using GLUs to condition models is common. This is often implemented as follows:
`input = someConvOp1(input)
condition=someConvOp2(condition)
input += condition
out = GLU(out)`

From my understanding this means that the sigmoid function is applied to half of the input and half of the condition and then applied as a gate to the other halves of the input and condition (in both of these cases the tensors have been added together).
This seems weird to me, of course it works but wouldn't a more efficient implementation be to apply the sigmoid to the condition in full then apply this gate to the full input?

Does this make sense or am I missing something?

Thanks in advance.

**Comment by hazdzz** [comment]
> You're also confusing the input to the convolution or linear layer with the input to the gated linear unit. The input to the convolution (**E**) is not split. The input to the gated linear unit is split (that's **A** and **B**). GLU implements:
> 
> 
> 
> ```
> 
> H0 = A⊗σ(B) 
> 
> ```
> 
> 
> 
> Where **A** and **B** are the two halves of the input to the gated linear unit (output of convolution).

Why not Linear(a) * Sigmoid(Linear(b))? The PyTorch implementation is not the same as original paper purposesd.

**Comment by shinxg** [comment]
> > You're also confusing the input to the convolution or linear layer with the input to the gated linear unit. The input to the convolution (**E**) is not split. The input to the gated linear unit is split (that's **A** and **B**). GLU implements:
> > ```
> > 
> > H0 = A⊗σ(B) 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Where **A** and **B** are the two halves of the input to the gated linear unit (output of convolution).
> 
> Why not Linear(a) * Sigmoid(Linear(b))? The PyTorch implementation is not the same as original paper purposesd.

Same question here.

**Comment by colesbury** [comment]
> Why not Linear(a) ⊗ Sigmoid(Linear(b))? The PyTorch implementation is not the same as original paper purposesd.

It's identical. Convolution is a linear operator (or affine if you include the bias) -- it's distributive. Mathematically, it does not matter if you perform two convolutions, **W** * **X** and  **V** * **X** or a single convolution **[W V]** * **X** with the weights concatenated. It's typically more computationally efficient to perform a single convolution operation using the concatenated weights (and split the output) than two smaller operations.
